### PR TITLE
inference/unify output buffer management

### PR DESCRIPTION
**Pull request by Superjomn**
- Make `PaddleBuf` possible to manage the buffer

### CODE REVIEW

File: paddle/contrib/inference/paddle_inference_api.cc [hunk 1]
```diff
@@ -13,3 +13,52 @@ See the License for the specific language governing permissions and
 limitations under the License. */
 
 #include "paddle/contrib/inference/paddle_inference_api.h"
+
+namespace paddle {
+
+PaddleBuf::PaddleBuf(PaddleBuf&& other)
```
Comment (panyx0718) [inline paddle/contrib/inference/paddle_inference_api.cc:19]: Is this used?

File: paddle/contrib/inference/paddle_inference_api.cc [hunk 2]
```diff
@@ -13,3 +13,52 @@ See the License for the specific language governing permissions and
...
+  return *this;
+}
+
+void PaddleBuf::Resize(size_t length) {
+  // Only the owned memory can be reset, the external memory can't be changed.
+  assert(memory_owned_);
+  Free();
+  data_ = new char[length];
+  length_ = length;
+  memory_owned_ = true;
```
Comment (panyx0718) [inline paddle/contrib/inference/paddle_inference_api.cc:46]: why assign again?

File: paddle/contrib/inference/paddle_inference_api_impl.cc [hunk 3]
```diff
@@ -241,10 +241,11 @@ bool NativePaddlePredictor::GetFetch(
     }
 
     outputs->at(i).shape = shape;
-    outputs->at(i).data.length = sizeof(float) * data.size();
-    outputs->at(i).data.data = malloc(outputs->at(i).data.length);
-    std::memcpy(
-        outputs->at(i).data.data, data.data(), outputs->at(i).data.length);
+    auto &buffer = outputs->at(i).data;
+    if (buffer.empty() || buffer.length() < sizeof(float) * data.size()) {
```
Comment (panyx0718) [inline paddle/contrib/inference/paddle_inference_api_impl.cc:245]: Maybe we don't want to allow ```buffer.length() > sizeof(float) * data.size()```? If that's the case, how does client know how many bytes to use?

File: paddle/contrib/inference/paddle_inference_api.cc [hunk 4]
```diff
@@ -13,3 +13,52 @@ See the License for the specific language governing permissions and
...
+  length_ = other.length_;
+  memory_owned_ = other.memory_owned_;
+  return *this;
+}
+
+void PaddleBuf::Resize(size_t length) {
+  // Only the owned memory can be reset, the external memory can't be changed.
+  assert(memory_owned_);
+  Free();
+  data_ = new char[length];
```
Comment (panyx0718) [inline paddle/contrib/inference/paddle_inference_api.cc]: If the ```length == length_```, can return early?

File: paddle/contrib/inference/paddle_inference_api.cc (hunk 1 above)
Comment (Superjomn) [reply paddle/contrib/inference/paddle_inference_api.cc:19]: It is used in https://github.com/PaddlePaddle/Paddle/pull/11569/files#diff-b931977496999381cc83b3a9ae0dbb17R45

File: paddle/contrib/inference/paddle_inference_api.cc (hunk 2 above)
Comment (Superjomn) [reply paddle/contrib/inference/paddle_inference_api.cc:46]: reset a new length.

File: paddle/contrib/inference/paddle_inference_api_impl.cc (hunk 3 above)
Comment (Superjomn) [reply paddle/contrib/inference/paddle_inference_api_impl.cc:245]: It is safe to `Resize` here.

If the client doesn't know how much to allocate or `Resize`, just init with a large enough memory, and do not resize it.

File: paddle/contrib/inference/paddle_inference_api.cc (hunk 2 above)
Comment (panyx0718) [reply paddle/contrib/inference/paddle_inference_api.cc:46]: I mean memory_owned_ is already true, as assert before.

**Review by panyx0718 — APPROVED** [review]

**Pull request merged**